### PR TITLE
Parallelize the hash join spill

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Use async source worker to parallelize the join spill processing
which is currently executed sequentially one operator at a time.
Note we spill all the hash build operators altogether for now.

### CODE REVIEW

File: velox/exec/HashBuild.cpp [hunk 1]
```diff
@@ -1140,18 +1140,63 @@ void HashBuild::reclaim(
...
+
+  std::vector<std::unique_ptr<SpillResult>> results;
+  results.reserve(spillTasks.size());
+  for (auto& spillTask : spillTasks) {
+    results.push_back(spillTask->move());
+  }
+
+  for (auto& result : results) {
+    if (result->error != nullptr) {
+      std::rethrow_exception(result->error);
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:1198]: Why not combine this loop with the previous one?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng What's the motivation for this change? I believe currently we do not support parallel hash table build with spilling, do we?

```
  // TODO: re-enable parallel join build with spilling triggered after
  // https://github.com/facebookincubator/velox/issues/3567 is fixed.
```

Also, I'm seeing 2 spill code paths: HashBuild::reclaim and HashBuild::runSpill. Do we need both? Do we need to modify HashBuild::runSpill as well?

**Comment by xiaoxmeng** [comment]
> @xiaoxmeng What's the motivation for this change? I believe currently we do not support parallel hash table build with spilling, do we?
> 
> ```
>   // TODO: re-enable parallel join build with spilling triggered after
>   // https://github.com/facebookincubator/velox/issues/3567 is fixed.
> ```
> 
> Also, I'm seeing 2 spill code paths: HashBuild::reclaim and HashBuild::runSpill. Do we need both? Do we need to modify HashBuild::runSpill as well?

This is to accelerate the spilling code path execution triggered by memory arbitration.

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 2]
```diff
@@ -2747,6 +2747,118 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
   }
 }
 
+DEBUG_ONLY_TEST_F(SharedArbitrationTest, joinBuildSpillError) {
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2750]: What does this test do?

Should we create separate test for different operators? This test file is already 3k lines long.

File: velox/exec/HashBuild.cpp [hunk 3]
```diff
@@ -1141,18 +1141,60 @@ void HashBuild::reclaim(
...
+      // guard and must not throw. The first error is already captured before
+      // this runs.
+      try {
+        spillTask->move();
+      } catch (const std::exception& e) {
+      }
+    }
+  });
+
+  std::vector<std::unique_ptr<SpillResult>> results;
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:1191]: Do we need this variable? We don't seem to use it.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2750]: This test is to verify the reclaim error is handled properly. We could separate them into separate file one per each operator. And move SharedArbitrator and SharedArbitratorTest into separate files.

File: velox/exec/tests/HashJoinTest.cpp [hunk 4]
```diff
@@ -23,6 +23,7 @@
 #include "velox/exec/HashBuild.h"
 #include "velox/exec/HashJoinBridge.h"
 #include "velox/exec/PlanNodeStats.h"
+#include "velox/exec/Spill.h"
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:26]: Why is this needed?

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2750]: > This test is to verify the reclaim error is handled properly.

Would you explain how it does that? It is a bit hard to read the test code.

File: velox/exec/tests/HashJoinTest.cpp (hunk 4 above)
Comment (xiaoxmeng) [reply velox/exec/tests/HashJoinTest.cpp]: I previously added test here. Removed.

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 5]
```diff
@@ -2747,6 +2747,118 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
...
+                      .hashJoin(
+                          {"t0"},
+                          {"u0"},
+                          PlanBuilder(planNodeIdGenerator)
+                              .values(vectors)
+                              .project({"c0 AS u0", "c1 AS u1", "c2 AS u2"})
+                              .planNode(),
+                          "",
+                          {"t1"},
+                          core::JoinType::kAnti)
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2833]: Is it important to use anti join here? Would inner or other type of join not allow to reproduce the necessary conditions?

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2750]: The test is to trigger spilling failure during hash join spill processing. I simplify it a bit.

File: velox/exec/HashBuild.cpp [hunk 6]
```diff
@@ -1141,18 +1141,58 @@ void HashBuild::reclaim(
...
+            buildOp->spiller_->spill();
+            VELOX_CHECK_EQ(buildOp->table_->numDistinct(), 0);
+            buildOp->table_->clear();
+            // Release the minimum reserved memory.
+            buildOp->pool()->release();
+            return std::make_unique<SpillResult>(nullptr);
+          } catch (const std::exception& e) {
+            LOG(ERROR) << "Spill from hash build pool "
+                       << buildOp->pool()->name() << " failed: " << e.what();
+            // The exception is passed to the caller thread which checks this in
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:1169]: I wonder if this comment is accurate. Somehow I don't see how this exception is making its way into advanceSpill. Would you clarify?

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 5 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2833]: Not important. Any query that generate a large amount of memory usage is fine.

Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2833]: I see. How does this test related to making HashBuild spilling run multi-threaded then? Just trying to understand. Would appreciate a clarification.

**Review by mbasmanova — COMMENTED** [review]
> Support parallel join build spill

We do not support parallel join build with spilling, which makes this title confusing. Wondering if there is a way to rephrase for clarity.

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 7]
```diff
@@ -2747,6 +2747,118 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
...
+        if (++numAppends != numDrivers) {
+          return;
+        }
+        VELOX_FAIL(errMsg);
+      }));
+
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Task::requestPauseLocked",
+      std::function<void(Task*)>(([&](Task* /*unused*/) {
+        LOG(ERROR) << "task pause wait";
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp]: Why LOG(ERROR) ?

File: velox/exec/HashBuild.cpp [hunk 8]
```diff
@@ -1141,18 +1141,58 @@ void HashBuild::reclaim(
 
   std::vector<Spiller::SpillableStats> spillableStats;
   spillableStats.reserve(spiller_->hashBits().numPartitions());
-  // TODO: consider to parallelize this disk spilling processing.
+
+  struct SpillResult {
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:1145]: We could. Will update

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 9]
```diff
@@ -2747,6 +2747,61 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
...
+                        {"u0"},
+                        PlanBuilder(planNodeIdGenerator)
+                            .values(vectors)
+                            .project({"c0 AS u0", "c1 AS u1", "c2 AS u2"})
+                            .planNode(),
+                        "",
+                        {"t1"},
+                        core::JoinType::kAnti)
+                    .planNode())
+          .assertResults(
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2797]: We do not except the query to complete successfully, hence, we won't get a chance to compare the results with DuckDB. Therefore, no need to setup DuckDB in the first place. Let's use copyResults and remove duckDbQueryRunner_ from AssertQueryBuilder's constructor as well as remove createDuckDbTable(vectors) code.

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 10]
```diff
@@ -2747,6 +2747,61 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
   }
 }
 
+TEST_F(SharedArbitrationTest, joinBuildSpillError) {
+  const int kMemoryCapacity = 32 << 20;
+  // Set a small memory capacity to trigger hash join spill.
+  setupMemory(kMemoryCapacity, 0);
+
+  const int numVectors = 16;
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2755]: Do we need 16 vectors? How did you choose this number, 16?

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 11]
```diff
@@ -2747,6 +2747,61 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
...
+
+  std::shared_ptr<core::QueryCtx> joinQueryCtx = newQueryCtx(kMemoryCapacity);
+
+  const int numDrivers = 4;
+  std::atomic<int> numAppends{0};
+  const std::string injectedErrorMsg("injected spillError");
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::SpillState::appendToPartition",
+      std::function<void(exec::SpillState*)>([&](exec::SpillState* state) {
+        if (++numAppends != numDrivers) {
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2769]: Does this logic make first 3 operators to succeed in spilling, but 4-th (last) fail?

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 12]
```diff
@@ -2747,6 +2747,61 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
   }
 }
 
+TEST_F(SharedArbitrationTest, joinBuildSpillError) {
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2750]: Would it make sense to run this test in "stress" mode to make sure it is not flaky?

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 9 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2797]: Removed.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 11 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2769]: I want to trigger the spilling error in the middle of the processing.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 10 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2755]: No particular reason but have tried the other number. Basically the test needs 32MB memory usage.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 12 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2750]: Sure.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 7 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp]: Removed. It is leftover from debugging.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 5 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp]: The change is covered by existing test. This additional test is to make sure that the parallel join path can catch spilling error properly which is new.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 11 above)
Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2769]: Got it, but why not trigger it for all operators, i.e. unconditionally?

Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2769]: Only the first exception is captured. I just want to throw in the middle

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 10 above)
Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2755]: > Basically the test needs 32MB memory usage.

Wondering why is this? Maybe add a comment to clarify? Other readers may have the same question. What happens if you use fewer vectors? Does the test fail?

Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2755]: Is this because kMemoryCapacity is 32MB? Would it make sense to add a helper function that takes number of bytes and produces vectors that use at least that much memory?

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 11 above)
Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2769]: I see... It is not guaranteed that appendToPartition is called only one per operator, is it? If that's not necessary the case then comparing numAppends with numDrivers is misleading. Would it be sufficient to just throw on second invocation?

File: velox/exec/HashBuild.cpp (hunk 6 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp]: The comment is not accurate. Updated.

Comment (mbasmanova) [reply velox/exec/HashBuild.cpp]: Thanks.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 9 above)
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2797]: Let's change assertResults to copyResults(pool()).

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 13]
```diff
@@ -2747,6 +2747,61 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, arbitrateMemoryFromOtherOperator) {
...
+        if (++numAppends != numDrivers) {
+          return;
+        }
+        VELOX_FAIL(injectedErrorMsg);
+      }));
+
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  VELOX_ASSERT_THROW(
+      AssertQueryBuilder(duckDbQueryRunner_)
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2778]: Let's remove duckDbQueryRunner_ from here.

Comment (mbasmanova) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2778]: You may need to build plan first:

```
auto plan = PlanBuilder...
AssertQueryBuilder(plan)...copyResults(pool());
```

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 9 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2797]: I updated internally

File: velox/exec/HashBuild.cpp [hunk 14]
```diff
@@ -1141,18 +1141,57 @@ void HashBuild::reclaim(
...
+            // The exception is captured and thrown by the caller.
+            return std::make_unique<SpillResult>(std::current_exception());
+          }
+        }));
+    if ((operators.size() > 1) && (spillExecutor != nullptr)) {
+      spillExecutor->add([source = spillTasks.back()]() { source->prepare(); });
+    }
+  }
+
+  auto syncGuard = folly::makeGuard([&]() {
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:1178]: I wonder if it makes sense to use Futures directly, instead of AsyncSource. This would remove a lot of redundant code like this guard and explicit catching and storing of exceptions.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 10 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2755]: That's good point that we could consider to have some utility to leverage estimate size for that. I will do that in a followup.

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 11 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2769]: One operator could call appendToPartition multiple times so there is not guarantee there. We could change to call a second time.

File: velox/exec/HashBuild.cpp (hunk 14 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:1178]: I will think about how to simplify AsyncSource in followup. Thanks!

**Review by mbasmanova — APPROVED** [review]
Thanks.
